stl/inc/xutility

```diff
@@ -4917,7 +4923,8 @@ template <class _Ty1, class _Ty2>
 _INLINE_VAR constexpr bool _Can_memcmp_elements<_Ty1*, _Ty2*, false> = is_same_v<remove_cv_t<_Ty1>, remove_cv_t<_Ty2>>;
 
 template <class _Elem1, class _Elem2>
-_INLINE_VAR constexpr bool _Can_memcmp_elements<_Elem1, _Elem2, false> = false;
+_INLINE_VAR constexpr bool _Can_memcmp_elements<_Elem1, _Elem2, false> =
+    is_same_v<_Elem1, bool>&& is_same_v<_Elem2, bool>;

```

Comment:
Why are we afraid to compare the object representations of `bool`s with those of non-`bool` objects of integral type with the same size?

Reply:
I guess:
```C++
char a[] = {2,2,2};
bool b[] = {true, true, true}
```
would `memcmp` as `false`, whereas we expect `true`, no?

Reply:
The problem is the concern about the representations other than 0 and 1. `bit_cast<bool>(char(-128)) == signed char(-128)` is `true` but `bit_cast<bool>(char(-128)) == unsigned char(-128)` is `false` because of integral promotion. If we don't care about the other representations we can `memcmp` between `bool` and non-`bool`.

> I guess:
> 
> ```c++
> char a[] = {2,2,2};
> bool b[] = {true, true, true}
> ```
> 
> would `memcmp` as `false`, whereas we expect `true`, no?

`char(2) == true` is `false` because of integral promotion (`char(2)` is promoted to `int(2)` and `true` is promoted to `1`).

Reply:
> I guess:
> 
> ```c++
> char a[] = {2,2,2};
> bool b[] = {true, true, true}
> ```
> 
> would `memcmp` as `false`, whereas we expect `true`, no?

`true` promotes to `int{1}` and `char{2}` promotes to `int{2}`, so `char{2} == true` is evaluated as `2 == 1` after the Usual Arithmetic Conversions, which evaluates to `false`.

Reply:
> The problem is the concern about the representations other than 0 and 1. `bit_cast<bool>(char(-128)) == signed char(-128)` is `true`

`bit_cast<bool>(char(x))` has undefined behavior if `char(x)` is any value other than `char{0}` or `char{1}` per [[[bit.cast]/2](http://eel.is/c++draft/bit.cast#2) "For the result and each object created within it, if there is no value of the object’s type corresponding to the value representation produced, the behavior is undefined."

Reply:
OK